File: worker/internal/worker/result_event.go
// Package worker holds types and logic for the KernelQ worker plane.
//
// This file defines the JSON event shape workers will publish when a job
// attempt finishes. Result events travel on kernelq.jobs.results so the
// Python control plane can update Postgres and drive retry policy.
package worker

import (
	"encoding/json"
	"fmt"
	"strings"
)

// ResultTopic is the Kafka topic where workers publish execution outcomes.
const ResultTopic = "kernelq.jobs.results"

// workerResultEventType is the only event_type value allowed on ResultTopic.
const workerResultEventType = "job.result"

// WorkerResultEvent is the on-wire contract for reporting job outcomes.
//
// Workers produce these after execution; the control plane consumes them to
// move jobs toward succeeded, failed/retry_scheduled, or dead_lettered.
type WorkerResultEvent struct {
	EventType string          `json:"event_type"`
	JobID     string          `json:"job_id"`
	Status    ExecutionStatus `json:"status"`
	Message   string          `json:"message"`
	Worker    string          `json:"worker"`
}

// NewWorkerResultEvent builds a result event from a job id, execution outcome,
// and worker identity. It does not validate—call Validate() or ToJSON() before
// publishing.
func NewWorkerResultEvent(jobID string, result ExecutionResult, worker string) WorkerResultEvent {
	return WorkerResultEvent{
		EventType: workerResultEventType,
		JobID:     jobID,
		Status:    result.Status,
		Message:   result.Message,
		Worker:    worker,
	}
}

// Validate checks required fields before we publish to kernelq.jobs.results.
func (e WorkerResultEvent) Validate() error {
	if e.EventType != workerResultEventType {
		return fmt.Errorf("event_type must be %q, got %q", workerResultEventType, e.EventType)
	}

	if strings.TrimSpace(e.JobID) == "" {
		return fmt.Errorf("job_id must not be blank")
	}

	// Reuse ExecutionResult rules: status must be a known constant; message may be blank.
	outcome := ExecutionResult{Status: e.Status, Message: e.Message}
	if err := outcome.Validate(); err != nil {
		return err
	}

	if strings.TrimSpace(e.Worker) == "" {
		return fmt.Errorf("worker must not be blank")
	}

	return nil
}

// ToJSON encodes the event as a JSON string after validation.
func (e WorkerResultEvent) ToJSON() (string, error) {
	if err := e.Validate(); err != nil {
		return "", err
	}

	data, err := json.Marshal(e)
	if err != nil {
		return "", err
	}

	return string(data), nil
}

File: control_plane/kernelq/result_event.py
"""
Worker result events consumed by the Python control plane.

Workers publish JSON to ``kernelq.jobs.results`` when a job attempt finishes.
The control plane will eventually read these events, validate them, and drive
Postgres state transitions (succeeded, retry, dead letter).
"""

from __future__ import annotations

import json
from dataclasses import dataclass

# Kafka topic where Go workers publish execution outcomes.
RESULT_TOPIC = "kernelq.jobs.results"

# Only this event_type is allowed on RESULT_TOPIC (matches the Go worker contract).
WORKER_RESULT_EVENT_TYPE = "job.result"

# Allowed ``status`` values on WorkerResultEvent (matches worker ExecutionStatus).
ALLOWED_STATUSES = frozenset(
    {
        "succeeded",
        "retryable_failure",
        "terminal_failure",
    }
)


@dataclass
class WorkerResultEvent:
    """
    On-wire shape for a job execution outcome reported by a worker.

    Field names match the JSON keys workers publish to Kafka.
    """

    event_type: str
    job_id: str
    status: str
    message: str
    worker: str

    def validate(self) -> None:
        """
        Check required fields before we trust this event for Postgres updates.

        Raises ValueError if anything is invalid. ``message`` may be blank.
        """
        if self.event_type != WORKER_RESULT_EVENT_TYPE:
            raise ValueError(
                f"event_type must be {WORKER_RESULT_EVENT_TYPE!r}, got {self.event_type!r}"
            )

        if not self.job_id.strip():
            raise ValueError("job_id must not be blank")

        if self.status not in ALLOWED_STATUSES:
            allowed = ", ".join(sorted(ALLOWED_STATUSES))
            raise ValueError(f"status must be one of: {allowed}, got {self.status!r}")

        if not self.worker.strip():
            raise ValueError("worker must not be blank")

    def to_dict(self) -> dict:
        """Convert to a plain dict with JSON field names."""
        return {
            "event_type": self.event_type,
            "job_id": self.job_id,
            "status": self.status,
            "message": self.message,
            "worker": self.worker,
        }

    def to_json(self) -> str:
        """Validate, then encode as a JSON string."""
        self.validate()
        return json.dumps(self.to_dict())


def parse_worker_result_event(data: bytes | str) -> WorkerResultEvent:
    """
    Parse and validate a worker result event from Kafka (or a test fixture).

    Args:
        data: Raw JSON as ``bytes`` or ``str``.

    Returns:
        A validated WorkerResultEvent.

    Raises:
        ValueError: Invalid JSON or invalid event fields.
    """
    if isinstance(data, bytes):
        try:
            text = data.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise ValueError("result event must be valid UTF-8") from exc
    else:
        text = data

    try:
        payload = json.loads(text)
    except json.JSONDecodeError as exc:
        raise ValueError("invalid JSON for worker result event") from exc

    if not isinstance(payload, dict):
        raise ValueError("worker result event must be a JSON object")

    event = WorkerResultEvent(
        event_type=str(payload.get("event_type", "")),
        job_id=str(payload.get("job_id", "")),
        status=str(payload.get("status", "")),
        message=str(payload.get("message", "")),
        worker=str(payload.get("worker", "")),
    )
    event.validate()
    return event
